New more natural colors for dark style

westnordost opened the pull request:
Inspired by https://github.com/openstreetmap/openstreetmap-website/issues/2332 (see https://issviewer.com/) and a little by Helium314's high contrast style for his fork.

### Goals:
- Improve readability by using more different hues (i.e. not everything is blue)
- Improve compatibility with overlays since overlays color roads and thus the road color in the background map should better be near black, white or at least be monochrome
- Use more natural colors (like in light mode) to help recognizing what is what

Click the images to see them full-size.

### z9
<img src="https://user-images.githubusercontent.com/4661658/203387580-0a06a0e0-3db8-43f9-90a3-36d1650e3b87.png"><img src="https://user-images.githubusercontent.com/4661658/203387563-bddb5d68-6b3f-4d49-a05f-cb304593d813.png"><img src="https://user-images.githubusercontent.com/4661658/203387588-c90e0fb1-b1f1-4b4f-ad4b-1c66e44aa8cd.png">

### z13
![dark-natural-13](https://user-images.githubusercontent.com/4661658/203388308-8168e70b-5b5f-4086-b3d5-43e6f25a8262.png)
![dark-city-13](https://user-images.githubusercontent.com/4661658/203388327-7122a5d6-d1d8-4512-a531-7d3911ac8f81.png)
![light-13](https://user-images.githubusercontent.com/4661658/203388346-9109dbb8-1c7b-432e-a669-9c5e888e65ff.png)

### z16
![dark-natural-16](https://user-images.githubusercontent.com/4661658/203388789-aba648d4-86d6-44d4-84fa-e2f1e5227483.png)
![dark-city-16](https://user-images.githubusercontent.com/4661658/203388805-84a8883e-1211-4015-a8c6-aedb1ae08b8e.png)
![light-16](https://user-images.githubusercontent.com/4661658/203388813-581c2bb4-176f-4a09-b6ea-aa5cda944fda.png)

### z19
![dark-natural-19](https://user-images.githubusercontent.com/4661658/203388855-fdbab9c5-4cfe-4577-b5fd-ff4294048120.png)
![dark-city-19](https://user-images.githubusercontent.com/4661658/203388860-20e078e0-85dd-4236-a6a9-f21e1c897cd0.png)
![light-19](https://user-images.githubusercontent.com/4661658/203388871-7363bd37-5e41-42b2-99d8-8a6561318280.png)

### Observations

- This natural dark style looks a little busy because the brightness for the different landuses varies more than for the light style for example. Making them all darker will make the roads less visible
- In the city dark style, the roads stood out more for being bright against a dark background
- The city dark style looks "cooler" on low zooms due to the high contrast between landmass and water, however this is not very useful

westnordost committed:
New more natural colors for dark style

westnordost committed:
less (hue) contrast etc

mnalis commented:
On first look, few comments:
- the most important are Z19, as users will probably spend most time there, so I'd optimize for that case
- in Z19 pictures, I'd actually prefer if the "useless" (non-road) elements which takes a lot of space (like landuses, buildings, rivers, parks, pitches etc.) were the ones which  were the darkest, and the roads the ones which are brighter.
- however in your next comment you show you tried with brighter roads, but:
  - that is too bright IMHO
  - have you really changed just the road brightness there? It looks to me like other elements are brighter too. It could be optical illusion, though.
- in any case I'd try to make less important elements (especially the ones like to be taking bigger areas) darker even darker then in examples above. They're not that important, and if one clicks on them for quest, they'd be visibly outlined anyway. 
 
I'd have to give it some more attention, but not tonight (probably should try it on the prone with brightness tuned down to get more realistic result).

Helium314 commented:
Some impressions / comparing with current dark theme, from just panning the map on my phone (during daytime, mostly high zoom)
* contrast is reduced (I needed to increase brightness to see as clearly as with current dark theme)
  * this is worse when there are (many) quests on screen, because pins and dots are much brighter than everything else
* the center lines of pedestrian roads (#119) imo look more confusing
* in cities, the colors somewhat give the impression that the area between houses and roads is a sidewalk... not sure if this is intended, but I like it
* roads and paths stand out much less when zooming out a bit, especially for paths this starts rather early (probably due to low contrast between road outline and residential landuse)
* one way arrows stand out too much

matkoniecz commented:
I always disable night mode so I am not the best person to review it, but for

> This is how it would look like with bright roads. Looks a bit busy IMO...

minor business increase seems worth it to see road structure, especially in presence of quests

westnordost commented:
> I always disable night mode

Why though? I do the same, but maybe it is because the night mode just doesn't look good.

For me, the main dealbreaker is that (even with the background not being completely black), the quest forms and dialogs kind of melt into the background and I find this really disrupting. Plus, I like much more the light style of the map.

westnordost commented:
> contrast is reduced (I needed to increase brightness to see as clearly as with current dark theme) 

With current dark theme, you mean your dark theme, or the blueish night theme?

westnordost commented:
So from the comments, I take away that the contrast between roads and rest is too low.

This means that the roads should likely be brighter than the landuses because the other option would be to leave the roads at black and increase the brightness of everything else.
If the roads are brighter than the landuses, the landuses could (optionally) be made even more dark, also for better contrast towards the blue water.

westnordost commented:
Just a short note, average brightness of 
light style: 85%
(new) dark style: 29%
(old) dark style: 43%

westnordost committed:
make darker

westnordost commented:
So, made everything darker except for the roads which are brighter now

![image](https://user-images.githubusercontent.com/4661658/203992193-e64a1f33-913e-46bc-b5ec-5f33a8d2bd87.png)
![image](https://user-images.githubusercontent.com/4661658/203992395-a6e78afd-f1b7-4168-8441-f82548d1bfda.png)
![image](https://user-images.githubusercontent.com/4661658/203992531-785c18c8-78ef-4001-ac1a-30ee872a83f7.png)

westnordost commented:
So what do you think?

mnalis commented on streetcomplete-dark-style.yaml:
```diff
@@ -13,32 +13,29 @@ import:
+    water_color: [0.106, 0.122, 0.243, 1.00]
+    green_color: [0.137, 0.145, 0.125, 1.00]
+    forest_color: [0.067, 0.110, 0.078, 1.00]
+    town_color: [0.087, 0.090, 0.090, 1.00]
```
```suggestion
    town_color: [0.0, 0.0, 0.0, 1.00]
```

I find empty space in cities to be completely black nicer and less crowded especially on high zoom levels (and it does not change display much on lower zooms):
![small_Screenshot_20221129_022556_de westnordost streetcomplete mn debug](https://user-images.githubusercontent.com/156656/204417771-7b6c6757-98f5-4d6e-813c-93cf803c3131.jpg) ![small_Screenshot_20221129_023931_de westnordost streetcomplete mn debug](https://user-images.githubusercontent.com/156656/204417776-90f17fe0-9b49-49ae-a63b-e887137c7a69.jpg)

mnalis replied:
(BTW it shows more noticeable on real phone; screenshots on desktop do not do it justice).

westnordost replied:
Completely black for landuses? But then, there is no diference in color between the "background" and e.g.  `landuse=residential`.

No, I think the issue here is that the contrast between the buildings and the background is too low in the left picture. In general, I think I overdid it a bit with the decrease in brightness.

After all, even in "dark mode" on the desktop PC, you have white text on black background. Not dark grey text on black background. 

And as people want to use the dark theme also in the sun (@Helium314 , ...), it would make sense if it had a little more contrast, too.

So maybe I should try out:
- lighter streets and names
- do something about the contrast between buildings and background
- slightly brighter landuses etc

mnalis reviewed (changes requested):
Other the suggested changing of town color to black, the map looks rather good to me ([tried](https://github.com/mnalis/StreetComplete/actions/runs/3570146675) it on real smartphone too).

One other related issue, but I'm not sure where it is defined:

Could we make backdrop of quest icons and small quest circles (as well as background of buttons / GPS pointer) some darker color then **intensive white** (maybe darker grey)? It's **ruining** the mood here (especially if there is more then a few of them).

![small_Screenshot_20221129_025738_de westnordost streetcomplete mn debug](https://user-images.githubusercontent.com/156656/204419840-b6b85b9f-0b5d-4186-8763-855bb6b18620.jpg)

westnordost commented:
Hm true, it is "ruining the mood". However, changing the color of the icons is not possible / would need to be implemented (so that two different files are generated, one for dark and one for light mode). 

The sprite sheet is created in `TangramPinsSpriteSheet` in case you want to look around. Basically to enable that there are two different files means that you have to save the version like in `Prefs.PIN_SPRITES_VERSION_LIGHT` and `Prefs.PIN_SPRITES_VERSION_DARK` or something.

Anyway, as written above...  the contrast is a bit extreme. I think "the rest" and the pins should not be that far away from each other, brightness-wise. Even if the pins were a white outline on a black pin, the quest icons themselves are all rather bight.

westnordost committed:
revision 3

westnordost commented:
Increased brightness of road (+railway + highway) color a lot and the rest a little

![Screenshot 2022-12-08 at 01 31 20](https://user-images.githubusercontent.com/4661658/206327167-ab5ceece-f22a-4f73-a577-92133a1b6c30.png)
![Screenshot 2022-12-08 at 01 32 50](https://user-images.githubusercontent.com/4661658/206327178-862d2e8d-e2dc-4422-8d55-7a0ff403eae2.png)
![Screenshot 2022-12-08 at 01 33 23](https://user-images.githubusercontent.com/4661658/206327181-c4eb345b-32f6-4e49-9f4b-55e3d474852e.png)

westnordost commented:
Comparison rev 1, rev 2, rev 3

<table>
<tr><th width="300">rev 1</th><th width="300">rev 2</th><th width="300">rev 3</th></tr>
<tr>
<td><img src="https://user-images.githubusercontent.com/4661658/203388855-fdbab9c5-4cfe-4577-b5fd-ff4294048120.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/203992395-a6e78afd-f1b7-4168-8441-f82548d1bfda.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/206327181-c4eb345b-32f6-4e49-9f4b-55e3d474852e.png"/></td>
</tr>
<tr>
<td><img src="https://user-images.githubusercontent.com/4661658/203388789-aba648d4-86d6-44d4-84fa-e2f1e5227483.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/203992193-e64a1f33-913e-46bc-b5ec-5f33a8d2bd87.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/206327178-862d2e8d-e2dc-4422-8d55-7a0ff403eae2.png"/></td>
</tr>
<tr>
<td><img src="https://user-images.githubusercontent.com/4661658/203388308-8168e70b-5b5f-4086-b3d5-43e6f25a8262.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/203992531-785c18c8-78ef-4001-ac1a-30ee872a83f7.png"/></td>
<td><img src="https://user-images.githubusercontent.com/4661658/206327167-ab5ceece-f22a-4f73-a577-92133a1b6c30.png"/></td>
</tr>
</table>

Maybe the text in rev3 should actually be black fill with white stroke (then, it could be made even a little bit brighter).
The buildings merge somewhat with the background in rev3, not sure if this is a bad thing.

Helium314 commented:
rev3 definitely works better inside SC than the previous one. (I created a branch with all 3 revisions selectable, plus my own ugly dark style, for more convenient in-app comparison)
I think the building outlines could be a bit brighter, this also helps against _buildings merge somewhat with the background_.

> color difference between water and railway

Colors are similar, but I still find them easily discernable.

Helium314 commented:
Oh, looks like I forgot the link to the branch...
https://github.com/Helium314/StreetComplete/tree/dark_compare

westnordost committed:
rev4

better visible buildings, slightly brighter roads, black text with white stroke

westnordost committed:
apply alpha fix

westnordost commented:
So, rev4 is pretty much the same as rev3 but:
- buildings are more red, i.e. better visible
- roads are slightly brighter (even more bright makes mid-level zoom really look ugly IMO, too high contrast)
- the text is now black with white stroke

westnordost committed:
add old dark style as "nightblue" style

westnordost committed:
rename

Helium314 commented:
> even more bright makes mid-level zoom really look ugly

I agree, it doesn't look too good already.

> the text is now black with white stroke

I find that considerably harder to read in some cases (depends a lot on what is behind the text)

I also prefer darker colors for town and earth, they also makes buildings a bit more visible.

----
Regarding the bright UI: https://github.com/Helium314/StreetComplete/tree/dark_compare2 has dark buttons and gray dots / pins (`global.pin_dot_color`), though a lazy implementation with unwanted side effect of darker quest icons.
(you'll need to try for yourself, as github claims my screenshot is empty...)

westnordost commented:
Hm, I took it out in the sunshine, and it is really all quite difficult to see. The roads look a bit dim but okay, all the rest is (still) barely visible. It is definitely worse than the current dark mode where both buildings and roads are ok visible even in sunlight.

At this point, I am not sure if both a quite dark background and enough contrast to see everything important well in sunlight is achievable. Does anyone actually use dark mode in the sunlight?

Helium314 commented:
Usability in sunlight was the reason I created my own dark style, which I find it quite usable any time (haven't used anything else in a long time).

I guess relatively bright outlines are a way of getting clearly visible buildings and roads, but that is a good step towards ugly

mnalis commented:
Yeah, I also have concerns that it is possible to have a coloring theme that is ***at the same time***:

- easy on eyes (and battery!) in the dark / **low-light surroundings**
- has high enough contrast in **bright daylight**

I would think one would need 3 themes:
- Dark mode for use in dark (low-contast, dilluted colors, no bright elements) - like `rev2` above which I quite like
- Dark mode for use in bright sunlight (high-contrast outlines, but otherwise mostly black to prevent battery overuse/overheating) - perhaps something like @Helium314 high-contrast mode. Such high-contrast might also help people with degraded vision (i.e. "accessibility feature").
- Bright mode (for people who prefer look of that, and don't care about overheating / battery use). (or, **_if_** there absolutely must only be two only themes, this would be the one to cull out -- but it would be sad as it actually looks rather good, especially for new users, so I would prefer 3 themes solution)

westnordost commented:
By the way, I tested the dark theme during night yesterday and I must say that even in the night, it is not well visible and also worse than the current dark mode. How could that be? I guess it is up to the fact that screen brightness automatically gets tuned down somewhat in low brightness environment. Especially so for the newer OLED displays (which I expect most current phones now employ), which are able to achieve much more dark colors.

So, I am not sure if this PR should even be followed up further. If it shall, then I guess I would need to turn up the overall brightness up a notch somewhat further. (This will make people like mnalis unhappy, but then again, people like mnalis use probably Helium314 high contrast style anyway.)

Helium314 commented:
Maybe users are happy with the current dark theme anyway? I don't remember any complaints except https://github.com/streetcomplete/StreetComplete/issues/3378

> I would need to turn up the overall brightness up a notch somewhat further, quite probably beyond rev1

The main issue I have with rev1 (and the others to some extent) is that the dark backgrounds seem needlessly bright imo.
With dark areas darker and bright things/areas/texts brighter there would be more contrast (obviously) and at the same time the white UI would stand out much less.

westnordost commented:
Okay, let's stash this for now. I will not delete the branch, but close the PR.
